### PR TITLE
Enable frame pointers in RelWithDebInfo build type

### PR DESCRIPTION
**Pull request by bjacob**
This makes profilers more useful, at least Linux's "perf".
`perf record -g` defaults to relying on frame pointers. There is a non-default option `perf record --call-graph=dwarf`, to rely on debug info instead. It produces inferior results and has a discoverability issue anyway.

### CODE REVIEW

**Comment by benvanik** [comment]
Not against frame pointers, but changing global cmake options is sketchy as it impacts anyone integrating IREE into their project. Can you scope this to our options instead? We should only be setting this on iree_runtime_cc_library/binary.

**Comment by bjacob** [comment]
good point, i'll do that tomorrow.

**Comment by bjacob** [comment]
Done, PTAL.

**Review by jpienaar — APPROVED** [review]
LGTM, fits my mental model of being able to profile properly if deb info is enabled.

File: build_tools/cmake/iree_copts.cmake
Comment (jpienaar) [inline build_tools/cmake/iree_copts.cmake]: Actually, is this related? Or would we want even other options here still?

Comment (bjacob) [reply build_tools/cmake/iree_copts.cmake]: It was originally related as I was initially adding the new code in this part of the file, so I removed this unneeded placeholder.  I didn't bother to restore it, it seemed to have outlived its purpose.

Comment (bjacob) [reply build_tools/cmake/iree_copts.cmake]: I'll wait for another approval from either @GMNGeoffrey, @ScottTodd or @pzread before going ahead with the merge.

File: build_tools/cmake/iree_copts.cmake
Comment (ScottTodd) [inline build_tools/cmake/iree_copts.cmake]: Build types are case insensitive, so this check should use something other than `STREQUAL` (e.g. convert to uppercase then check)

https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html
https://cmake.org/cmake/help/latest/manual/cmake-buildsystem.7.html#build-configurations
https://cmake.org/cmake/help/latest/manual/cmake-buildsystem.7.html#case-sensitivity
https://gitlab.kitware.com/cmake/cmake/-/issues/22591

File: build_tools/cmake/iree_copts.cmake
Comment (ScottTodd) [reply build_tools/cmake/iree_copts.cmake]: I'd keep those lines.

File: build_tools/cmake/iree_copts.cmake
Comment (bjacob) [reply build_tools/cmake/iree_copts.cmake]: Thanks for the tip. Done, copied the TOUPPER line from `function(iree_fix_ndebug)` in our root CMakeLists.txt.

File: build_tools/cmake/iree_copts.cmake
Comment (ScottTodd) [inline build_tools/cmake/iree_copts.cmake]: almost ;)

```suggestion
string(TOUPPER "${CMAKE_BUILD_TYPE}" _UPPERCASE_CMAKE_BUILD_TYPE)
if (IREE_DEV_MODE OR (_UPPERCASE_CMAKE_BUILD_TYPE STREQUAL "RELWITHDEBINFO"))
```

Comment (bjacob) [reply build_tools/cmake/iree_copts.cmake]: lol, thanks, good catch (sorry, that shows i hadn't even tested this locally)

**Review by ScottTodd — APPROVED** [review]

**Pull request merged**
